qcodes/instrument_drivers/devices.py

```diff
@@ -1,6 +1,190 @@
+
+
+class Role(enum.Enum):
+    GAIN = enum.auto()
+    DIVISION = enum.auto()
+
+
+class ParameterScaler(Parameter):
+    """
+    Parameter Scaler
+
+    To be used when you use a physical voltage divider or an amplifier to set
+    or get a quantity.
+
+    Initialize the parameter by passing the parameter to be measured/set
+    and the value of the division OR the gain.
+
+    The scaling value can be either a scalar value or a Qcodes Parameter.
+
+    The parameter scaler acts a your original parameter, but will set the right
+    value, and store the gain/division in the metadata.
+
+    Examples:
+        Resistive voltage divider
+        >>> vd = ParameterScaler(dac.chan0, division = 10)
+
+        Voltage multiplier
+        >>> vb = ParameterScaler(dac.chan0, gain = 30, name = 'Vb')
+
+        Transimpedance amplifier
+        >>> Id = ParameterScaler(multimeter.amplitude, division = 1e6, name = 'Id', unit = 'A')
+
+    Args:
+        output: Physical Parameter that need conversion
+        division: the division value
+        gain: the gain value
+        label: label of this parameter, by default uses 'output' label
+            but attaches _amplified or _attenuated depending if gain
+            or division has been specified
+        name: name of this parameter, by default uses 'output' name
+            but attaches _amplified or _attenuated depending if gain
+            or division has been specified
+        unit: resulting unit. It uses the one of 'output' by default
+    """
+
+    def __init__(self,
+                 output: Parameter,
+                 division: Union[int, float, Parameter] = None,
+                 gain: Union[int, float, Parameter] = None,
+                 name: str=None,
+                 label: str=None,
+                 unit: str=None) -> None:
+        self._wrapper_param = output
+
+        # Set the role, either as divider or amplifier
+        # Raise an error if nothing is specified
+        is_divider = division is not None
+        is_amplifier = gain is not None
+
+        if not (is_divider ^ is_amplifier):
+            raise ValueError('Provide only division OR gain')
+
+        if is_divider:
+            self.role = Role.DIVISION
+            self._multiplier = division
+        elif is_amplifier:
+            self.role = Role.GAIN
+            self._multiplier = gain
+
+        # Set the name
+        if name:
+            self.name = name
+        else:
+            if self.role == Role.DIVISION:
+                self.name = "{}_attenuated".format(self._wrapper_param.name)
+            elif self.role == Role.GAIN:
+                self.name = "{}_amplified".format(self._wrapper_param.name)
+
+        # Set label
+        if label:
+            self.label = label
+        elif name:
+            self.label = name
+        else:
+            if self.role == Role.DIVISION:
+                self.label = "{}_attenuated".format(self._wrapper_param.label)
```

Comment:
The amount of times decisions are made based on the ‘role’ means that this should be implemented via factory so that the code is more maintainable. Is it clear what I am suggesting, or do you need help/support with this?

Reply:
Yes, I can't really understand this. Do you have some example/doc?

Reply:
:) There is a design pattern called "factory", it allows you to create similar objects while keeping the specifics of the objects inside them. For example, now the ParameterScaler can be a divider, an amplifier, and also knows how to choose its role. The class becomes too big, complex, with repetitive code (the `if role == ..` stuff), not easily testable as you have to remember about all the possible things the class can do. Instead, an more maintainable design is to have a Divider class, an Amplifier class, and a "factory" class that provides the interface for creating them.

You can read more [on the wiki](https://en.wikipedia.org/wiki/Factory_method_pattern) or take a look at [a somewhat close example in Python](https://krzysztofzuraw.com/blog/2016/factory-pattern-python.html).

So my proposal would be to:
* define Divider class (perhaps private so that it cannot be directly instantiated) that only divides
* define Amplifier class (perhaps private so that it cannot be directly instantiated) that only amplifies
* keep the ParameterScaler class as an "interface" that users will use to create Dividers or Amplifiers based on the key-value argument (`gain=` or `division=`)
* refactor existing tests so that they correspond to these classes.

@YakBizzarro , in case you don't have time or you find it troublesome to implement the factory, would you allow me to do it? (this will mean that I will push some commits to you branch)